### DISCUSSION TITLE
EnsureEmailIsVerified Middleware Bug?

### DISCUSSION THREAD

**Discussion by iAdil** [laravel/framework#46149] [Q&A]
The middleware called "EnsureEmailIsVerified" verifies if the user model is an instance of "MustVerifyEmail". However, it won't function properly for the "Illuminate\Foundation\Auth\User" class because it uses a trait.

Is it a bug?

**Comment by ahmett** [comment]
I see nothing is wrong with the core logic. Your confusion is probably about naming. `EnsureEmailIsVerified` middleware checks whether your user model is an instance of `Illuminate\Contracts\Auth\MustVerifyEmail`

If you want your users required to verify their emails, you have to implement the `Illuminate\Contracts\Auth\MustVerifyEmail` interface in your user class (which is `App\Models\User` by default).

**For a better understanding of the structure:** 

Application's user model `App\Models\User` extends `Illuminate\Foundation\Auth\User` and implements `Illuminate\Contracts\Auth\MustVerifyEmail`

_Example user model:_

```php
<?php
 
namespace App\Models;
 
use Illuminate\Contracts\Auth\MustVerifyEmail;
use Illuminate\Foundation\Auth\User as Authenticatable;
 
class User extends Authenticatable implements MustVerifyEmail
{
    // ...
}
```

Under the hood, `Illuminate\Foundation\Auth\User` uses the `Illuminate\Auth\MustVerifyEmail` trait to fulfill the requirements of `Illuminate\Contracts\Auth\MustVerifyEmail`

```php
<?php

namespace Illuminate\Foundation\Auth;

use Illuminate\Auth\MustVerifyEmail;

class User
{
    use MustVerifyEmail;
}
```

**_Note:_** I deleted the unnecessary details in the core code above to make everything more precise with context.

_You can also check:_ https://laravel.com/docs/10.x/verification#model-preparation
